### PR TITLE
Add Product Image Gallery 

### LINKED ISSUE #8233: Create Product Image Gallery block
We want to create a block for the product images on the single product page that wraps the current behavior rendered by PHP.

| Property | Value |
|---------|---------|
| Block title | Product Image Gallery |
| Block description | TBD |
| Block slug | TBD |
| Block file location | TBD |
| Block class location | `src/BlockTypes/ProductImageGallery.php` |
| Scope | Site Editor |
| Equivalent template file | [`plugins/woocommerce/templates/single-product/product-image.php`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/templates/single-product/product-image.php) |
| Equivalent template function | [`woocommerce_show_product_images`](https://github.com/woocommerce/woocommerce/blob/25f051cad273bf709b127d162ab43f437d774c42/plugins/woocommerce/includes/wc-template-functions.php#L1549) |
| Style | TBD |
| Visibility | Woocommerce Core |

### Screenshots
image

### PR DESCRIPTION
**Pull request by gigitux**
This PR adds the Product Image Gallery block. It is possible to add this block only on the Single Product Image Gallery thanks to [this](https://github.com/woocommerce/woocommerce-blocks/blob/1d63441d7a14bf1dd4eabf2c83f2480e965ca467/assets/js/atomic/utils/register-block-single-product-template.ts).

| Block Details | Preview Editor Side | Frontend  |
|---------------|---------------------|-----------|
|image|image|image|

### Design Feedback
@vivialice could give me feedback about:
1. Icon to use.
2. Description.
3. Preview on the editor side.

### Notes for testing this PR.

Because of #8314, the styles of this block are not loaded. To avoid this issue, add the class `product` to the main div that wraps the entire page.

Fixes #8233

#### User Facing Testing

1. Via Site Editor, edit the Single Product Template.
2. Add the `Product Image Gallery`. Save it.
3. On the frontend side, open a product page, check that the Product Image Gallery is loaded, and show the right images.
4. Check that the block can be added only on the Single Product Template

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Add Product Image Gallery block.

### CODE REVIEW

File: assets/js/atomic/blocks/product-elements/product-image-gallery/edit.tsx
Comment (imanish003) [inline assets/js/atomic/blocks/product-elements/product-image-gallery/edit.tsx]: Is there any specific reason we are adding width & height of `img` here as well as inside `editor.scss` ?

File: assets/js/atomic/blocks/product-elements/product-image-gallery/editor.scss
Comment (imanish003) [inline assets/js/atomic/blocks/product-elements/product-image-gallery/editor.scss]: I am nitpicking here but maybe we can improve the naming of this class. Some of alternatives could be:
- `wc-block-editor-product-gallery__alternative-images`
- `wc-block-editor-product-gallery__other-images`

If it doesn't break the frontend.

File: assets/js/atomic/utils/register-block-single-product-template.ts
Comment (imanish003) [inline assets/js/atomic/utils/register-block-single-product-template.ts:18]: Maybe we can add 2nd argument to subscribe function i.e. storeNameOrDescriptor parameter, the listener function will be called only on updates on that one specific registered store. In our case, I believe it would be `core/edit-site`. 
What do you think?

File: src/BlockTypes/ProductImageGallery.php
Comment (imanish003) [inline src/BlockTypes/ProductImageGallery.php:39]: Is there any specific reason behind defining `$product` as global here?

**Review by imanish003 — APPROVED** [review]
Hi @gigitux, I have tested the PR & everything is working as expected after adding `product` class. Excellent work 🙌🏻 💪🏻

I have left a few minor suggestions, but those shouldn't block this PR. Therefore, I am pre-approving the PR.

File: assets/js/atomic/blocks/product-elements/product-image-gallery/edit.tsx
Comment (gigitux) [reply assets/js/atomic/blocks/product-elements/product-image-gallery/edit.tsx]: I forgot to remove the inline style :| 
I did the fix via [795a698](https://github.com/woocommerce/woocommerce-blocks/pull/8235/commits/795a6980d6e084b183b7499f163689df18e4bad2) commit.

File: assets/js/atomic/blocks/product-elements/product-image-gallery/editor.scss
Comment (gigitux) [reply assets/js/atomic/blocks/product-elements/product-image-gallery/editor.scss]: Thanks for the suggestion! I picked `wc-block-editor-product-gallery__other-images`. [795a698](https://github.com/woocommerce/woocommerce-blocks/pull/8235/commits/795a6980d6e084b183b7499f163689df18e4bad2)

File: assets/js/atomic/utils/register-block-single-product-template.ts
Comment (gigitux) [reply assets/js/atomic/utils/register-block-single-product-template.ts:18]: Maybe I don't understand what do you mean, but the `subscribe` function receive only the callback as parameter.

File: src/BlockTypes/ProductImageGallery.php
Comment (gigitux) [reply src/BlockTypes/ProductImageGallery.php:39]: Yep. It is necessary, otherwise, `woocommerce_show_product_images()` doesn't work.

**Comment by gigitux** [comment]
Thanks for the review and feedback! I addressed your comments.
Also, I added the custom class support via bfd2e98f615e894ce84fb3dde1d6d2b6862ac603 commit.

**Comment by gigitux** [comment]
Before merging the PR, I prefer to wait for feedback from @vivialice.

**Comment by gigitux** [comment]
@imanish003 JFYI I removed the save function because it is not necessary for the server-side blocks 👍 [e526eff](https://github.com/woocommerce/woocommerce-blocks/pull/8235/commits/e526eff91a4bd6e955e62793f5b18c259305e5ce)

File: assets/js/atomic/utils/register-block-single-product-template.ts
Comment (dinhtungdu) [inline assets/js/atomic/utils/register-block-single-product-template.ts:18]: Could we use debounce here to make this trigger less frequent?

File: src/BlockTypes/ProductImageGallery.php
Comment (imanish003) [reply src/BlockTypes/ProductImageGallery.php:39]: Got it. Thanks for explaining it to me 🤝

File: assets/js/atomic/utils/register-block-single-product-template.ts
Comment (imanish003) [reply assets/js/atomic/utils/register-block-single-product-template.ts:18]: Sorry for not being clear.

I checked [this doc](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-data/#subscribe) and it looks like there is 2nd argument that could be passed. Here is the text from the link:

```
Given a listener function, the function will be called any time the state value
of one of the registered stores has changed. If you specify the optional
storeNameOrDescriptor parameter, the listener function will be called only
on updates on that one specific registered store.

This function returns an unsubscribe function used to stop the subscription.

Usage

import { subscribe } from '@wordpress/data';

const unsubscribe = subscribe( () => {
    // You could use this opportunity to test whether the derived result of a
    // selector has subsequently changed as the result of a state update.
} );

// Later, if necessary...
unsubscribe();

Parameters

    listener Function: Callback function.
    storeNameOrDescriptor string|StoreDescriptor?: Optional store name.
```

Please let me know if I am missing something.

Comment (gigitux) [reply assets/js/atomic/utils/register-block-single-product-template.ts:18]: Oh, nice! I updated the PR and also I created a PR on DefinitelyTyped repo https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64177

Comment (gigitux) [reply assets/js/atomic/utils/register-block-single-product-template.ts:18]: Mmm, I would avoid a debounce here to avoid any kind of race condition.

Comment (dinhtungdu) [reply assets/js/atomic/utils/register-block-single-product-template.ts:18]: I'm curious what can be the race condition and its drawback. The motivation behind my suggestion is to keep the editor performant.

Comment (imanish003) [reply assets/js/atomic/utils/register-block-single-product-template.ts:18]: Thanks for creating a PR on DefinitelyTyped repo 🙌🏻

**Comment by vivialice** [comment]
Hey @gigitux ! 👋 

For the **block icon**, I think it's best to use the same `gallery` block as core. It is in line with other feedback we have received, so that works best until we do more with this block.

I took a look at **the preview**, and we should align with the greys used in wireframe layouts in core. However, I did notice there are two styles being used to represent images: 

_Screenshot of one of the Wireframe pattern image placeholders:_
Screenshot 2023-02-02 at 14 09 56

_Screenshot of the Image block placeholder:_ 
Screenshot 2023-02-02 at 14 10 31

I think it's best to align with the Wireframe style. I have created a mockup: 

Screenshot 2023-02-02 at 14 12 26

Do you need an image or svg to implement the above? 

For the **block description**, let's use `Display a product's images.`

**Comment by gigitux** [comment]
Thanks @vivialice for the feedback! 🙇 

>Do you need an image or svg to implement the above?

Could you share the SVG?

**Comment by vivialice** [comment]
Sure thing! Here you go @gigitux 

[Download](https://cloudup.com/coNTr6RFUdh)

File: assets/js/atomic/utils/register-block-single-product-template.ts
Comment (gigitux) [reply assets/js/atomic/utils/register-block-single-product-template.ts:18]: Mmm, for example, the user is editing another template, but the function isn't called due to the debounce and the block is still visible in the inserter.

Comment (dinhtungdu) [reply assets/js/atomic/utils/register-block-single-product-template.ts:18]: > but the function isn't called due to the debounce

I don't think we will have this issue, as we just want to debounce them (by 100ms, for example), the subscription still triggers, but less frequently. So when the user switches to another template, the subscription may not trigger right away, but still fast enough to register/unregister the block before user interaction.

We don't have to do it now btw, we can go with the current PR and come back to this later when we see it affect the editor performance.

**Comment by gigitux** [comment]
> Sure thing! Here you go @gigitux
> 
> [Download](https://cloudup.com/coNTr6RFUdh)

Ops! Sorry, I didn't specify. Could you share only this piece of the image?
image

**Comment by gigitux** [comment]
Via 8aee84125143d2ecf3eaf3af1b74ca449b034cb2 I updated the package `@types/wordpress__data` and removed the comment that I added after [this discussion](https://github.com/woocommerce/woocommerce-blocks/pull/8235#discussion_r1091730765).

**Comment by gigitux** [comment]
@vivialice I updated the image, this is the preview:

image

I'm going to merge the PR, feel free to comment here so that I can create a follow-up PR addressing your feedback! 🙇

**Pull request merged**
